File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/liveride/LiveRideService.java
package net.cyclestreets.liveride;

import android.Manifest;
import android.util.Log;
import net.cyclestreets.util.Logging;
import net.cyclestreets.util.PermissionsKt;
import org.osmdroid.util.GeoPoint;

import net.cyclestreets.routing.Journey;
import net.cyclestreets.routing.Route;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.Binder;
import android.os.Bundle;
import android.os.IBinder;

public class LiveRideService extends Service implements LocationListener
{
  private IBinder binder_;
  private LocationManager locationManager_;
  private Location lastLocation_;
  private LiveRideState stage_;
  private static final int UPDATE_DISTANCE = 5;  // metres
  private static final int UPDATE_TIME = 500;    // milliseconds
  private static final String TAG = Logging.getTag(LiveRideService.class);

  @Override
  public void onCreate() {
    binder_ = new Binding();
    locationManager_ = (LocationManager)getSystemService(Context.LOCATION_SERVICE);
    stage_ = LiveRideStateKt.stoppedState(this);
  }

  @Override
  public int onStartCommand(final Intent intent, final int flags, final int startId) {
    return Service.START_NOT_STICKY;
  }

  @Override
  public void onDestroy() {
    stopRiding();
    super.onDestroy();
  }

  @Override
  public IBinder onBind(final Intent intent) {
    return binder_;
  }

  public void startRiding() {
    if (!stage_.isStopped())
      return;

    if (!PermissionsKt.hasPermission(this, Manifest.permission.ACCESS_FINE_LOCATION)) {
      // Should be unreachable, but we're being defensive
      Log.w(TAG, "Location permission is not granted.  Bail out.");
      return;
    }

    stage_ = LiveRideStateKt.initialState(this);
    locationManager_.requestLocationUpdates(LocationManager.GPS_PROVIDER, UPDATE_TIME, UPDATE_DISTANCE, this);
  }

  public void stopRiding() {
    if (stage_.isStopped())
      return;
    stage_.getTts().stop();
    stage_.getTts().shutdown();
    stage_ = LiveRideStateKt.stoppedState(this);
    locationManager_.removeUpdates(this);
  }

  public boolean areRiding() {
    return stage_.arePedalling();
  }

  public Location lastLocation() {
    return lastLocation_;
  }

  public class Binding extends Binder  {
    private LiveRideService service() { return LiveRideService.this; }
    public void startRiding() { service().startRiding(); }
    public void stopRiding() { service().stopRiding(); }
    public boolean areRiding() { return service().areRiding(); }
    public String stage() { return stage_.getClass().getSimpleName(); }
    public Location lastLocation() { return service().lastLocation(); }
  }

  // ///////////////////////////////////////////////
  // location listener
  @Override
  public void onLocationChanged(final Location location) {
    if (!Route.available()) {
      stopRiding();
      return;
    }

    lastLocation_ = location;

    final GeoPoint whereIam = new GeoPoint(location);
    final float accuracy = location.hasAccuracy() ? location.getAccuracy() : 2;

    final Journey journey = Route.journey();

    stage_ = stage_.update(journey, whereIam, (int)accuracy);
  }

  @Override
  public void onProviderDisabled(String arg0) { }
  @Override
  public void onProviderEnabled(String arg0) { }
  @Override
  public void onStatusChanged(String arg0, int arg1, Bundle arg2) { }
}

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/liveride/LiveRideService.kt
package net.cyclestreets.liveride

import android.Manifest
import android.annotation.SuppressLint
import android.util.Log
import net.cyclestreets.util.Logging
import net.cyclestreets.util.*
import org.osmdroid.util.GeoPoint

import net.cyclestreets.routing.Route
import android.app.Service
import android.content.Context
import android.content.Intent
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.os.Binder
import android.os.Bundle
import android.os.IBinder
import android.speech.tts.TextToSpeech
import android.speech.tts.TextToSpeech.ERROR
import android.speech.tts.TextToSpeech.SUCCESS

private const val UPDATE_DISTANCE = 5f  // metres
private const val UPDATE_TIME = 500L    // milliseconds
private val TAG = Logging.getTag(LiveRideService::class.java)

class LiveRideService : Service(), LocationListener, TextToSpeech.OnInitListener {
    private lateinit var binder: IBinder
    private lateinit var locationManager: LocationManager
    private var stage: LiveRideState? = null
    private var lastLocation: Location? = null

    override fun onCreate() {
        binder = Binding()
        locationManager = getSystemService(Context.LOCATION_SERVICE) as LocationManager
        stage = Stopped(this)
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        return Service.START_NOT_STICKY
    }

    override fun onDestroy() {
        stopRiding()
        super.onDestroy()
    }

    override fun onBind(intent: Intent): IBinder? {
        return binder
    }

    @SuppressLint("MissingPermission") // We handle this with the hasPermission() check
    fun startRiding() {
        if (!stage!!.isStopped())
            return

        if (!hasPermission(this, Manifest.permission.ACCESS_FINE_LOCATION)) {
            // Should be unreachable, but we're being defensive
            Log.w(TAG, "Location permission is not granted.  Bail out.")
            return
        }

        val tts = TextToSpeech(this, this)
        tts.setOnUtteranceProgressListener(AudioFocuser(this))
        stage = LiveRideStart(this, tts).setServiceForeground(this)
        locationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, UPDATE_TIME, UPDATE_DISTANCE, this)
        Log.d(TAG, "startRiding")
    }

    fun stopRiding() {
        if (stage!!.isStopped())
            return
        stage!!.tts!!.stop()
        stage!!.tts!!.shutdown()
        stage = Stopped(this)
        locationManager.removeUpdates(this)
        Log.d(TAG, "stopRiding")
    }

    inner class Binding : Binder() {
        private fun service(): LiveRideService { return this@LiveRideService }
        fun startRiding() { service().startRiding() }
        fun stopRiding() { service().stopRiding() }
        fun areRiding(): Boolean { return service().stage!!.arePedalling() }
        fun lastLocation(): Location? { return service().lastLocation }
    }

    // TextToSpeech init listener
    @SuppressLint("MissingPermission")
    override fun onInit(status: Int) {
        Log.i(TAG, "TextToSpeech init returned $status ($SUCCESS=success, $ERROR=error)")
        locationManager.getLastKnownLocation(LocationManager.GPS_PROVIDER)?.let { onLocationChanged(it) }
    }

    // Location listener
    override fun onLocationChanged(location: Location) {
        if (!Route.routeAvailable()) {
            stopRiding()
            return
        }

        lastLocation = location

        val whereIam = GeoPoint(location)
        val accuracy: Int = if (location.hasAccuracy()) location.accuracy.toInt() else 2
        stage = stage!!.update(Route.journey(), whereIam, accuracy)
    }

    override fun onProviderDisabled(arg0: String) {}
    override fun onProviderEnabled(arg0: String) {}
    override fun onStatusChanged(arg0: String, arg1: Int, arg2: Bundle) {}
}
